Kernel/FileSystem/File.cpp

```diff
@@ -24,6 +24,7 @@ bool File::unref() const
 {
     if (deref_base())
         return false;
+    const_cast<File&>(*this).before_removing();
```

Comment:
Hmm. Like I mentioned on discord, if this virtual we're overriding did all the magic that RefCounted (instead of RefCountedBase) does, we could use the magic `will_be_destroyed()` pattern instead of inventing the new  before_removing() function.

```c++
    bool File::unref() const
    {
        auto new_ref_count = deref_base();
        if (new_ref_count == 0) {
            call_will_be_destroyed_if_present(static_cast<const T*>(this));
            delete static_cast<const T*>(this);
            return true;
        } else if (new_ref_count == 1) {
            call_one_ref_left_if_present(static_cast<const T*>(this));
        }
        return false;
    }
```

Reply:
Wait I'm re-reading this comment and I don't know what I'm talking about. you've already implemented the pattern as designed, just with a different name .we're not in a template context, this is the concrete File class.